Домашнє завдання до теми "Основи роботи з даними у Pandas"
В цьому домашньому завданні продовжуємо працювати з набором даних supermarket_sales.csv.

Імпортуйте бібліотеку pandas та зчитайте дані у змінну df типу pandas.DataFrame.

In [176]:
import pandas as pd

In [177]:
import os
os.listdir('../data')

['.ipynb_checkpoints',
 'sample_submission.csv',
 'supermarket_sales.csv',
 'train.csv']

In [178]:
data_path = ('../data/supermarket_sales.csv')
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [179]:
df.head(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


In [180]:
df.tail(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [181]:
df.loc[615:619]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
615,509-29-3912,A,Yangon,Member,Female,Sports and travel,38.72,9,17.4240,365.9040,3/20/2019,12:24,Ewallet,348.48,4.761905,17.4240,4.2
616,327-40-9673,B,Mandalay,Member,Male,Sports and travel,72.60,6,21.7800,457.3800,1/13/2019,19:51,Cash,435.60,4.761905,21.7800,6.9
617,840-19-2096,C,Naypyitaw,Member,Male,Electronic accessories,87.91,5,21.9775,461.5275,3/14/2019,18:10,Ewallet,439.55,4.761905,21.9775,4.4
618,828-46-6863,A,Yangon,Member,Male,Food and beverages,98.53,6,29.5590,620.7390,1/23/2019,11:22,Credit card,591.18,4.761905,29.5590,4.0
619,641-96-3695,C,Naypyitaw,Member,Female,Fashion accessories,43.46,6,13.0380,273.7980,2/7/2019,17:55,Ewallet,260.76,4.761905,13.0380,8.5


3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.
   
    

In [184]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

    

In [185]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

75-му персентилю відповідає значення = 22.445250. 
Це означає, що 75% всіх операцій продажів з 3х філій магазину мають валовий дохід який менше або дорівнює цьому значеннню. Всі інші транзакції (25%) приносять дохід вищий за 22.445250. 

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [186]:
df['City'].value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [187]:
city_counts = df['City'].value_counts()

In [188]:
invoices_percentage_by_city = city_counts/city_counts.sum()

In [189]:
print(invoices_percentage_by_city)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: count, dtype: float64


7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [190]:
df_selected_cols = df[['City', 'gross income']]
df_selected_cols 

,City,gross income
0,Yangon,26.1415
1,Naypyitaw,3.8200
2,Yangon,16.2155
3,Yangon,23.2880
4,Yangon,30.2085
...,...,...
995,Naypyitaw,2.0175
996,Mandalay,48.6900
997,Yangon,1.5920
998,Yangon,3.2910


In [191]:
#7.1
yangon_mean = df.query("City == 'Yangon'")['gross income'].mean()
mandalay_mean = df.query("City == 'Mandalay'")['gross income'].mean()

print(f"Gross income в Yangon становить: {yangon_mean} і воно менше за gross income в Mandalay, який = {mandalay_mean}")

Gross income в Yangon становить: 14.874001470588235 і воно менше за gross income в Mandalay, який = 15.232024096385542


7.2
1. **Гіпотеза №1 - спосіб оплати**:
   
В Mandalay покупці частіше розраховуються безготівковими способами(credit card/ewallet), що стимулює робити більш дорогі покупки.

Перевіряємо гіпотезу за наявними даними:
розраховано середній gross income для кожного способу оплати в цих містах. 

Маємо наступні результати:
- Credit card: Mandalay = 16.31 vs Yangon = 15.15
- Ewallet: Mandalay = 14.12 vs Yangon = 14.86
- Cash: Mandalay = 15.30 vs Yangon = 14.62

Необхідні додаткові дані про загальну кількість/частку транзакцій за кожним методом оплати в кожному місті.

Рекомендації: мотивувати покупців у Yangon та Mandalay частіше використовувати кредитні картки (через кешбек, партнерські акції з банками, вигідну розстрочку), оскільки цей сегмент дає найбільший середній чек.

2. **Гіпотеза №2 - вплив програми лояльності**:
   
В Mandalay більше покупців є учасниками програми лояльності (Member), які здійснюють більші за обсягом покупки порівняно зі звичайними відвідувачами (Normal).

Перевіряємо гіпотезу за наявними даними:
порівнюємо середній **gross income** для типів **Member** та **Normal** у кожному місті.

Маємо наступні результати:
Учасники **Member** купують на більшу суму в обох містах: Mandalay = 15.50, Yangon = 15.29.
Клієнти **Normal** витрачають менше: Mandalay = 14.97, Yangon = 14.47.

Гіпотеза про те, що **Member** купують більше за **Normal**, підтвердилась.
Проте це не пояснює розрив між містами, оскільки середній чек у Mandalay вищий в обох категоріях покупців.

Необхідні додаткові дані про відсоток покупок від Member у загальній кількості чеків кожної філії, а також частота повернення (LTV) цих клієнтів.

Рекомендації: з огляду на те, що статус Member гарантовано підвищує середній чек, потрібно стимулювати перехід покупців з Normal у Member (картки лояльності, бонуси за реєстрацію).
Дослідити, чому саме у Mandalay навіть категорія Normal витрачає більше (14.97), ніж у Yangon (14.47). Можливо, у Mandalay загалом вищі ціни або інша викладка товарів.

3. **Гіпотеза №3 - перевірка обсягу покупки та ціни (Quantity та Unit price)**:

В Mandalay вищий gross income, тому що покупці беруть більшу кількість товарів за один візит (Quantity) або купують дорожчі позиції (Unit price).

In [192]:
# Кількість товарів у чеку
yangon_qty = df.query("City == 'Yangon'")['Quantity'].mean()
mandalay_qty = df.query("City == 'Mandalay'")['Quantity'].mean()

print(f"Середня кількість товарів: Yangon = {yangon_qty:.2f}, Mandalay = {mandalay_qty:.2f}")

# Ціна за одиницю товару
yangon_price = df.query("City == 'Yangon'")['Unit price'].mean()
mandalay_price = df.query("City == 'Mandalay'")['Unit price'].mean()

print(f"Середня ціна товару: Yangon = {yangon_price:.2f}, Mandalay = {mandalay_price:.2f}")

Середня кількість товарів: Yangon = 5.47, Mandalay = 5.48
Середня ціна товару: Yangon = 54.78, Mandalay = 55.66


8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [193]:
#option 1 
df['gross margin percentage'].unique()

array([4.76190476])

In [194]:
#option 2
df['gross margin percentage'].nunique()

1

**Відповідь:** у всіх всіх інвойсах однакова маржа

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [195]:
yangon_invoice_distribution = (df.query('City == "Yangon"')['Product line'].value_counts(normalize=True)*100).round(2)
mandalay_invoice_distribution = (df.query('City == "Mandalay"')['Product line'].value_counts(normalize=True)*100).round(2)

print("Yangon:\n", yangon_invoice_distribution), 

print("\n\n")

print("Mandalay: \n", mandalay_invoice_distribution)

Yangon:
 Product line
Home and lifestyle        19.12
Electronic accessories    17.65
Sports and travel         17.35
Food and beverages        17.06
Fashion accessories       15.00
Health and beauty         13.82
Name: proportion, dtype: float64



Mandalay: 
 Product line
Fashion accessories       18.67
Sports and travel         18.67
Electronic accessories    16.57
Health and beauty         15.96
Food and beverages        15.06
Home and lifestyle        15.06
Name: proportion, dtype: float64
